server/monitor.go

```diff
@@ -0,0 +1,504 @@
+	s.monMu.RLock()
+	numSubs := s.numSubs
+	s.monMu.RUnlock()
+	now := time.Now()
+	serverz := &Serverz{
+		ClusterID:     s.info.ClusterID,
+		ServerID:      s.serverID,
+		Version:       VERSION,
+		GoVersion:     runtime.Version(),
+		State:         state.String(),
+		Now:           now,
+		Start:         s.startTime,
+		Uptime:        myUptime(now.Sub(s.startTime)),
+		Clients:       s.store.GetClientsCount(),
+		Channels:      numChannels,
+		Subscriptions: numSubs,
+		TotalMsgs:     count,
+		TotalBytes:    bytes,
+	}
+	sendResponse(w, r, serverz)
+}
+
+func myUptime(d time.Duration) string {
+	// Just use total seconds for uptime, and display days / years
+	tsecs := d / time.Second
+	tmins := tsecs / 60
+	thrs := tmins / 60
+	tdays := thrs / 24
+	tyrs := tdays / 365
+
+	if tyrs > 0 {
+		return fmt.Sprintf("%dy%dd%dh%dm%ds", tyrs, tdays%365, thrs%24, tmins%60, tsecs%60)
+	}
+	if tdays > 0 {
+		return fmt.Sprintf("%dd%dh%dm%ds", tdays, thrs%24, tmins%60, tsecs%60)
+	}
+	if thrs > 0 {
+		return fmt.Sprintf("%dh%dm%ds", thrs, tmins%60, tsecs%60)
+	}
+	if tmins > 0 {
+		return fmt.Sprintf("%dm%ds", tmins, tsecs%60)
+	}
+	return fmt.Sprintf("%ds", tsecs)
+}
+
+func (s *StanServer) handleStorez(w http.ResponseWriter, r *http.Request) {
+	count, bytes, _ := s.store.MsgsState(stores.AllChannels)
+	storez := &Storez{
+		ClusterID:  s.info.ClusterID,
+		ServerID:   s.serverID,
+		Now:        time.Now(),
+		Type:       s.store.Name(),
+		Limits:     *s.storeLimits,
+		TotalMsgs:  count,
+		TotalBytes: bytes,
+	}
+	sendResponse(w, r, storez)
+}
+
+type byClientID []*Clientz
+
+func (c byClientID) Len() int           { return len(c) }
+func (c byClientID) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+func (c byClientID) Less(i, j int) bool { return c[i].ID < c[j].ID }
+
+func (s *StanServer) handleClientsz(w http.ResponseWriter, r *http.Request) {
+	singleClient := r.URL.Query().Get("client")
+	subsOption, _ := strconv.Atoi(r.URL.Query().Get("subs"))
+	if singleClient != "" {
+		var clientz *Clientz
+		client := s.store.GetClient(singleClient)
+		if client != nil {
+			clientz = getMonitorClient(client, subsOption)
+		}
+		if clientz == nil {
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte(fmt.Sprintf("Client %s not found", singleClient)))
+			return
+		}
+		sendResponse(w, r, clientz)
+	} else {
+		offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))
+		if offset < 0 {
+			offset = 0
+		}
+		limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
+		if limit <= 0 {
+			limit = defaultMonitorListLimit
+		}
+		clients := s.store.GetClients()
+		totalClients := len(clients)
+		carr := make([]*Clientz, 0, totalClients)
+		for _, c := range clients {
+			cz := &Clientz{ID: c.ID}
+			carr = append(carr, cz)
+		}
+		sort.Sort(byClientID(carr))
+
+		minoff := offset
+		if minoff > totalClients {
+			minoff = totalClients
+		}
+		maxoff := offset + limit
+		if maxoff > totalClients {
+			maxoff = totalClients
+		}
+		carr = carr[minoff:maxoff]
```

Comment:
But at the end, since both store implementations maintain channels (same for clients) as a map, in order to return a subset, we would still need to create an array with all channels in order to sort and then apply the offset/limit and return a subset of this array.
Even with adding an internal number on a channel structure, we would need to store the channels in two maps, one per name, one per number and that could be used as the offset key. Problem is for clients (that can be removed) and will be for channels when we add the ability to remove them.
For channels, after a server restart, the numbering would not be the same since recovery will not guarantee the same ordering than when channels were first created. So it means that getting offset/limit between server restart would return inconsistent value. With current ordering based on name, it will still be ok.
What do you think?